Add limiter to Burgers executable

fmahebert opened the pull request:
Before this is fully ready to go, I think the following is needed:
- (by the group:) Discussion of how the limiter actions should be bundled into the actions list
- (by me:) Cleaning up of tags for initializing limiter

But I'm putting this PR in now so the code can be used as reference ASAP.

Commit order:
1. Add a step function shock solution for Burgers
2. Clarify shock speed in Burgers documentation
3. Enable LTS in ScalarWave executable
4. Disable LTS in Burgers executable
5. Add limiter to Burgers executable

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Add limiter to Burgers executable

fmahebert committed:
Disable LTS in Burgers executable

We want this executable to provide an example of using a limiter to
control a shock... and for now the limiter does not support local
time stepping.

fmahebert committed:
Add a step function shock solution for Burgers

fmahebert committed:
Clarify shock speed in Burgers documentation

fmahebert committed:
Enable LTS in ScalarWave executable

wthrowe commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -82,12 +86,23 @@ struct EvolutionMetavars {
+      // Triggering the limiting is likely beyond the intended scope of
+      // update_variables. Because limiting should be applied after time steps
+      // in both the evolution and the self-start procedure, adding the actions
+      // here does the job. Should discuss possibility of a better design.
```
I think what you have done is the right way to do this.

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/Burgers/Step.hpp:
```diff
@@ -0,0 +1,95 @@
+/// The shock propagates left-to-right, with profile
+/// \f$U(x, t) = U_R + (U_L - U_R) H(-(x - x_0 - v t))\f$.
+/// Here \f$U_L\f$ and \f$U_R\f$ are the left and right constant states,
+/// satisfying \f$U_L < U_R\f$; \f$H\f$ is the Heaviside function; \f$x_0\f$ is
```
`U_L > U_R`

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/Burgers/Step.hpp:
```diff
@@ -0,0 +1,95 @@
+  struct LeftValue {
+    using type = double;
+    static type lower_bound() { return 0.0; }
+    static type default_value() { return 1.0; }
```
The defaults here and for RightValue give an inverted shock, which will fail the assertion in the constructor.

fmahebert replied:
Whoops, thanks for catching this

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/Burgers/Step.hpp:
```diff
@@ -0,0 +1,95 @@
+  };
+  struct InitialPosition {
+    using type = double;
+    static type default_value() { return -0.5; }
```
Strange default.  I suggest either 0 or no default.

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/Burgers/Step.cpp:
```diff
@@ -0,0 +1,77 @@
+      right_value_(right_value),
+      initial_shock_position_(initial_shock_position) {
+  ASSERT(left_value > right_value,
+         "Shock solution expects left_value > right_value");
```
PARSE_ERROR (because checking input file values)

fmahebert replied:
Can you clarify what change you have in mind? It looks like PARSE_ERROR expects an OptionContext; how do I get this from the constructor?

wthrowe replied:
Add a `const OptionContext& context = {}` argument as the last constructor argument.  The option parser will detect this and pass one.

kidder commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -82,12 +86,23 @@ struct EvolutionMetavars {
   using component_list = tmpl::list<DgElementArray<
       EvolutionMetavars, dg::Actions::InitializeElement<1>,
       tmpl::flatten<tmpl::list<
+          // Optionally, one may wish to limit the initial data. This is not
```
I would just make an issue for this to examine later.  Once we have action lists per phase this will be easy to do.

kidder reviewed (commented):
you can squash changes

fmahebert commented:
Updated, squashed, rebased.

I opted to leave the initialization as is (i.e., specific to Minmod). I think it makes more sense to wait until we have at least 2 limiters before spending the time to generalize the initialization.

kidder commented:
@wthrowe @nilsdeppe this is ready

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Burgers/Test_Step.cpp:
```diff
@@ -0,0 +1,50 @@
+
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.AnalyticSolutions.Burgers.Step",
+                  "[PointwiseFunctions][Unit]") {
+  test_creation<Burgers::Solutions::Step>(
```
`test_creation` returns an object. You should test that it was set up correctly. All this is testing is that we can actually parse that string, but not that we parse it correctly.

fmahebert replied:
Done. Also added a new commit that adds similar checking of what was created by `test_creation` in several other tests that were missing this.

nilsdeppe reviewed (changes requested):
I'm happy for you to squash the changes.

fmahebert committed:
Make tests check the output of test_creation

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/WaveEquation/Test_PlaneWave.cpp:
```diff
@@ -157,12 +157,22 @@ void test_1d() {
+  const auto created_solution =
+      test_creation<ScalarWave::Solutions::PlaneWave<1>>(
+          "  WaveVector: [3.5]\n"
+          "  Center: [3.5]\n"
```
The only thing comparing these to explicitly constructed objects does is verifies that the factory and constructor agree on the order of the arguments, so make all the arguments different so that that is checked.

fmahebert replied:
Updated to use the same arguments as when constructing the solution object `pw` earlier in the test, which a) results in different arguments in the different slots, as requested, and b) simplifies the code by reusing `pw` in the comparison.

kidder commented:
can you doxygen document the MinMod options structs please

fmahebert commented:
@kidder Can you please clarify what you are looking for?
1. We do not use doxygen on most (if any at all?) of our option structs; why should this be different? Are we discussing requiring doxygen documentation for all option structs?
1. Would it be sufficient to point back to the documentation of Minmod? Or is the Minmod documentation unclear?

Regardless, I feel that improving documentation lies outside the scope of this PR, but I am happy to open another PR addressing your concerns.

fmahebert commented:
Rebased to address conflict with develop.

kidder commented:
Yes, the options should show up in the documentation, and can point to other parts of the documentation.  I am fine with an issue being filed.

moxcodes commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -161,6 +162,7 @@ static const std::vector<void (*)()> charm_init_node_funcs{
     &Parallel::register_derived_classes_with_charm<MathFunction<1>>,
     &Parallel::register_derived_classes_with_charm<
         StepChooser<metavariables::step_choosers>>,
+    &Parallel::register_derived_classes_with_charm<StepController>,
```
@wthrowe should this be removed from EvolveBurgers now that LTS is disabled there?

wthrowe replied:
The StepChooser and StepController registrations are unnecessary without LTS.  They should also be harmless, except for a (hopefully small?) increase in compile time.  In general we need to figure out a better way to manage this section.

Probably remove them for now to keep things clean.

fmahebert replied:
Okay, thanks for the pointers, @wthrowe!

moxcodes reviewed (commented):
Group code review with @fmahebert @mgiesler @geoffrey4444  and @markscheel

fmahebert commented:
Updated to remove unneeded registrations in Burgers executable as per Will's reply.

nilsdeppe reviewed (approved):
I restarted a GCC debug build that had a segfault for testing burgers executable input file.
